feat: fail validation for multiline cron jobs

rocketeerbkw opened the pull request:
Part of https://github.com/uselagoon/build-deploy-tool/issues/103.

The way YAML handles strings allows users to configure cron jobs with multiline commands. As detailed in the above issue, this can cause issues during builds. In addition to calling this out in the docs, this PR also adds a failing lint for this scenario.

I've added tests for all the multiline scenarios outlined at https://yaml-multiline.info/.

rocketeerbkw committed:
feat: fail validation for multiline cron jobs

rocketeerbkw committed:
fix: validate-config-map-json loading a deprecated ConfigMap

rocketeerbkw commented:
Updated to fix two issues:

1. For `build-deploy-tool` the cron job pipeline is `.lagoon.yml > shyaml > bash > yq > helm`. In `lagoon-linter` the cron job pipeline is `.lagoon.yml > go unmarshal`. I guess it shouldn't be surprising that the command strings end up in a different format between these two methods. Using `shyaml` causes  the strings to get trimmed, so I've updated the linter to also trim and added tests that match existing production cron jobs that were failing the first version of this PR.
2. The `validate-config-map-json` command was using a deprecated ConfigMap name and has been updated.

shreddedbacon reviewed (approved):
LGTM, will get @smlx to review too though to see if it meets the other requirement

smlx commented on internal/lagoonyml/required/lagoon.go:
```diff
@@ -14,6 +14,11 @@ type LagoonRoute struct {
 }
 
+type LagoonCronjob struct {
+	Name    string
```
This works, but the json unmarshaler will [search through permutations](https://go.dev/blog/json#decoding) of the field name to find the JSON field it's looking for. If you add a struct tag like `` `json:"name"` `` it shortcuts this behaviour and makes unmarshalling more predictable. See e.g. the `Environment` struct below.

rocketeerbkw replied:
I was copying the style of [`LagoonRoute`](https://github.com/uselagoon/lagoon-linter/blob/main/internal/lagoonyml/required/lagoon.go#L13) which doesn't have tags. What's the difference between the two? Or should the route also have tags?

smlx replied:
Ah I see the confusion - LagoonRoute is different because it has an `UnmarshalJSON` [method](https://github.com/uselagoon/lagoon-linter/blob/main/internal/lagoonyml/required/lagoon.go#L18). That means it implements the `json.Unmarshaler` [interface](https://pkg.go.dev/encoding/json#Unmarshaler), so the default `json.Unmarshal` logic will be skipped and the custom `UnmarshalJSON` logic will be used instead. It's just the default unmarshal logic that uses the struct tags.

rocketeerbkw replied:
TIL, thanks!

smlx reviewed (changes requested):
A few minor style things but looks good.

smlx commented:
The other thing we'll need to fix up is the documentation. Right now when the linter fails it [links to the Lagoon docs](https://github.com/uselagoon/build-deploy-tool/blob/main/legacy/build-deploy-docker-compose.sh#L218), but only to the route annotation docs. Maybe the linter should be a separate section in the docs?

rocketeerbkw commented on internal/lagoonyml/required/cronjobs.go:
```diff
@@ -0,0 +1,30 @@
+// newlines, and nil otherwise.
+func ValidateCronjob(c *LagoonCronjob) error {
+	if strings.Contains(strings.TrimSpace(c.Command), "\n") {
+		return fmt.Errorf("%q",
```
Good catch, thanks. This used to have a useful message, but I forgot to add it back after debugging the `trim` stuff.

rocketeerbkw committed:
chore: code review updates

rocketeerbkw commented:
Updated per the review. Also none of us caught that I added `invalid.6.lagoon.yml` but forgot to actually test it 🙈, so that's fixed now also.

I think the docs link is OK for now, the `build-deploy-tool` will validate cronjobs and error before lagoon-linter. Maybe we can update the structure so that each error can link to the relevant part of the `.lagoon.yml` docs when we move it over.

smlx reviewed (approved):
Nice catch with the missing invalid yaml test. LGTM.
